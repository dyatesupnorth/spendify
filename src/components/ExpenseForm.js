import React, { Component } from 'react'
import moment from 'moment';

import {SingleDatePicker} from 'react-dates';



const now = moment();
console.log('date: ', now.format('MMM Do, YYYY'));
export class ExpenseForm extends Component {
    constructor(props){
        super(props)
        this.state = {
            description: props.expense? props.expense.description : '',
            note : props.expense? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt : props.expense? moment(props.expense.createdAt) : moment(),
            calendarFocused : false,
            error: false
        }
    }
    
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>({description}));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    }
    onAmountChange = (e) =>{
        const amount = e.target.value;
        if (amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(()=>({amount}))
        } 
    }
    onDateChange = (createdAt) => {
        console.log('createdAt: ', createdAt);
        if (createdAt) {
            this.setState(() => ({createdAt}));    
        }
        

    }
    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}));
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please provide description and amount'}));

        }else{
            this.setState(() => ({error: false}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10)*100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note

            })
            console.log('submitted');
        }
    }
    render() {

        return (
                <div className="row">
                    <div className="col-sm-12  ml-md-auto">
                        {this.state.error}
                        <form onSubmit={this.onSubmit}>
                        <div className="form-group">
                            <label>Expense Description</label>
                            <input type="text"
                                className="form-control"
                                placeholder="Laptop, Coffee, etc"
                                autoFocus
                                value={this.state.description}
                                onChange={this.onDescriptionChange}
                                />
                        </div>
                        <div className="form-group">
                            <label>Amount</label>
                            <input 
                            className="form-control"
                            type="text"
                            placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}
                            />
                        </div>
                                <div className="form-group">
                                    <label>Choose Date</label>
                                    <br/>
                                    <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}

                            />
                                </div>

                            
                        <div className="form-group">
                            <label>Note (optional)</label>
                            <textarea className="form-control" placeholder="Add note for your expense (optional)" 
                            value={this.state.note} onChange={this.onNoteChange}>
                            </textarea>
                        </div>
                            

                            <button className="btn btn-primary">Save Expense</button>
                        </form>
                    </div>
                </div>
        )
  }
}

export default ExpenseForm
