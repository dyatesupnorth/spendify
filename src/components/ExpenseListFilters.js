import React from 'react'
import { connect } from 'react-redux'
import {DateRangePicker} from 'react-dates'
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
export class ExpenseListFilters extends React.Component{
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) => {
       this.setState({ calendarFocused })
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
        console.log(e.target.value);
    }
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        }else if(e.target.value === 'amount'){
            this.props.sortByAmount();
        }
    }
    render() {
      return (
        <div className="col-sm-12 col-md-4 p-3">
        <h4>Filter</h4>
        <hr/>
        <div className="form-group">
            <label>Name</label>
            <input className="form-control" type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
        </div>
        <div className="form-group">
            <label>Sort by</label>
            <select className="form-control" value={this.props.filters.sortBy} onChange={this.onSortChange}>
                <option value="date">Date
                </option>
                <option value="amount">Amount
                </option>
            </select>
        </div>
        <div className="form-group">
            <label>Select Date Range</label>
            <DateRangePicker
                startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
            />
        </div>
    </div>
      )
    }
};

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }   
} 
const mapDispatchToProps = (dispatch, ownProps) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate : () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
