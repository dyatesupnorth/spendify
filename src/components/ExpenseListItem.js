import React from "react";
import { removeExpense } from "../actions/expenses";
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
numeral.register('locale', 'en-gb', {
    delimiters: {
        thousands: ',',
        decimal: '.'
    },
    abbreviations: {
        thousand: 'th',
        million: 'm',
        billion: 'b',
        trillion: 't'
    },
    ordinal : function (number) {
        return number === 1 ? 'pound' : 'pounds';
    },
    currency: {
        symbol: 'Â£'
    }
});
numeral.locale('en-gb');
export const ExpenseListItem = ({id, description, amount,createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
        {numeral(amount / 100).format('$0,0.00')}
        - 
        {moment(createdAt).format('MMMM Do, YYYY')}</p>
    </div>
)

export default ExpenseListItem
