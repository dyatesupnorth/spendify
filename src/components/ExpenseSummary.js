import React from "react";
import {connect} from 'react-redux';
import selectExpenses from "../selectors/expenses";
import expensesTotal from "../selectors/expenses-total";
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


export const ExpenseSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return(
        <div>
        <p>Expense Summary</p>
        <p>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}.</p>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expensesTotal: expensesTotal(visibleExpenses),
        expenseCount:visibleExpenses.length
    }
}
export default connect(mapStateToProps)(ExpenseSummary);