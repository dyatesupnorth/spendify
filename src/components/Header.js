import React from 'react';
import { NavLink } from 'react-router-dom';
import {startLogout} from '../actions/auth';
import {connect} from 'react-redux';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem
 } from 'reactstrap';

export const Header = ({startLogout}) => (
  <div>
        <Navbar color="light" light expand="md">
          <NavbarBrand href="/">Spendify</NavbarBrand>
            <Nav className="navbar-nav ml-auto" navbar>
              <NavItem >
                <NavLink to="/dashboard" className="nav-link" activeClassName="active" >Go home</NavLink>
        
              </NavItem>
              <NavItem>
              <NavLink to="/create" className="nav-link" activeClassName="active">Create an Expense</NavLink>
                    </NavItem>
            
              <NavItem>
              <button className="btn btn-info" onClick={startLogout}>Logout</button>  

              </NavItem>
            </Nav>
        </Navbar>
      </div>
 
)
const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})
export default connect(undefined, mapDispatchToProps)(Header)
