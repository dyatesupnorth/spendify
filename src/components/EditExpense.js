import React from "react";
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { removeExpense, startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpense extends React.Component{
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.removeExpense({ id: this.props.expense.id })
        this.props.history.push('/')
    }
    render(){
        return(
            <div className="container">
            <ExpenseForm 
            expense={this.props.expense}
                onSubmit={this.onSubmit}
            
            />
            <hr/>
            <div className="row">
                <div className="col-sm-12">
                    <button className="btn btn-danger" onClick={this.onRemove}>Remove</button>

                </div>
            </div>
        
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch, props) => ({
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        removeExpense: (data) => dispatch(removeExpense(data))
})
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);