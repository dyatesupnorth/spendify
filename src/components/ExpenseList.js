import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div className="col-sm-12 col-md-8 p-3">
  <h4>My Expenses</h4>
  <hr/>
    {
      props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
          props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )
    }
  </div>
);  

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
