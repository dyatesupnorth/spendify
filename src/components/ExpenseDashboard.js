import React from "react";
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpenseSummary';

const ExpenseDashboard = () => (
    <div>
        <ExpensesSummary/>
        <div className="container">
            <div className="row">
            <ExpenseList />
            <ExpenseListFilters />
            </div>
        </div>
        
    </div>
)
export default ExpenseDashboard;