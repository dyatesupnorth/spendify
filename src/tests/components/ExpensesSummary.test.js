import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should correctly expenses summary with one expense', () => {
  const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={235} />)
  expect(wrapper).toMatchSnapshot();
})

test('should correctly expenses summary with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={4} expensesTotal={3432343} />)
    expect(wrapper).toMatchSnapshot();
})