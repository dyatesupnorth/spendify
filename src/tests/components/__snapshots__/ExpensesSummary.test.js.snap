// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly expenses summary with multiple expenses 1`] = `
<div>
  <p>
    Expense Summary
  </p>
  <p>
    Viewing 
    4
     
    expenses
     totalling 
    $34,323.43
    .
  </p>
</div>
`;

exports[`should correctly expenses summary with one expense 1`] = `
<div>
  <p>
    Expense Summary
  </p>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
    .
  </p>
</div>
`;
